### PR TITLE
[OC-1257] Set groups of entities compatible with reponse card 

### PR DESCRIPTION
**Pull request by quinarygio**
Release notes:

In Features:
[OC-1257] Set groups of entities compatible with response card. Added "entityAllowedToSendCard "boolean field to Entity.
NOTE: If an Entity has entityAllowedToSendCard=false and has child entities, than child entities will be checked recursively 

[OC-1257]: https://opfab.atlassian.net/browse/OC-1257

### CODE REVIEW

File: ui/main/src/app/services/entities.service.ts [hunk 1]
```diff
@@ -117,4 +117,25 @@ export class EntitiesService extends CachedCrudService implements OnDestroy {
     templateGateway.setEntityNames(entityNames);
   }
 
+  public static getEntitiesAllowedToRespond(selected: Entity[], allEntities: Entity[]) : Entity[] {
+    let allowed = new Array<Entity>();
+    selected.forEach(entity => {
+        console.log("Foreach entity " + entity.id);
```
Comment (freddidierRTE) [inline ui/main/src/app/services/entities.service.ts:123]: remove console.log

File: ui/main/src/app/services/entities.service.ts [hunk 2]
```diff
@@ -117,4 +117,25 @@ export class EntitiesService extends CachedCrudService implements OnDestroy {
...
+        if (entity.entityAllowedToSendCard) {
+            console.log("Push entity " + entity.id);
+            allowed.push(entity);
+        } else {
+          console.log("Get childs of entity " + entity.id);
+          const childsAllowed = EntitiesService.getEntitiesAllowedToRespond(EntitiesService.getChildEntities(entity, allEntities), allEntities);
+          console.log("Concat childs of entity " + entity.id + " size: " + childsAllowed.length);
```
Comment (freddidierRTE) [inline ui/main/src/app/services/entities.service.ts:130]: remove console.log

Comment (quinarygio) [reply ui/main/src/app/services/entities.service.ts]: Done

File: ui/main/src/app/services/entities.service.ts (hunk 1 above)
Comment (quinarygio) [reply ui/main/src/app/services/entities.service.ts]: Done

**Review by freddidierRTE — CHANGES_REQUESTED** [review]
1) Add or modify a karate test to set/get the new field entityAllowedToSendCard

2) Add in documentation src/docs/asciidoc/reference_doc/users_management.adoc (chapter entities) 

3) Add a goup of entity in the standard conf to faciliate test : in config/dev/users-dev.yml and in config/docker/users-docker.yml
--> 

 ```
 entities:
    - id: ENTITY1
      name: Control Room 1
      description: Control Room 1
      parents : ["ALLCONTROLROOMS"]
    - id: ENTITY2
      name: Control Room 2
      description: Control Room 2
      parents : ["ALLCONTROLROOMS"]
    - id: ENTITY3
      name: Control Room 3
      description: Control Room 3
      parents : ["ALLCONTROLROOMS"]
    - id: ALLCONTROLROOMS
      name: All Control Rooms 
      description: All Control Rooms
      entityAllowedToSendCard: false
    - id: ENTITY4
      name: IT Supervision Center 
      description: IT Supervision Center 


```
4) WIth the previous config when sending usercard "Example for new cards 2 /  question" to ALLCONTROLROOMS , it is not possible to respond for operator2 --> it shall be possible 

5) WIth the previous config when sending usercard "Example for new cards 2 /  question" to ENTITY1 (controlroom1) and ENTITY2(controlroom2) , when responding with operator2 , the respond is made on behalf of entity ALLCONTROLROOMS instead of ENTITY2
image

6) The list of answers in the card header should reflect the list of enitity Allowed to respond taking into account the entityAllowedToSendCard

image

File: ui/main/src/app/services/entities.service.spec.ts [hunk 3]
```diff
@@ -52,4 +53,43 @@ describe('EntitiesService', () => {
...
+        expect(allowedEntities[1].id).toBe('ENTITY3.1.2');
+        expect(allowedEntities[2].id).toBe('ENTITY3.2');
+      });
+      const req = httpMock.expectOne(`${environment.urls.entities}`);
+      expect(req.request.method).toBe('GET');
+      req.flush(listEntities);
+
+    });
+  });
+
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply ui/main/src/app/services/entities.service.spec.ts]: Done

File: ui/main/src/app/services/entities.service.spec.ts [hunk 4]
```diff
@@ -52,4 +53,43 @@ describe('EntitiesService', () => {
       req.flush(listEntities);
     });
   });
+  
+
+  describe('# test getEntitiesAllowedToRespond', () => {
+    it('should return an Entity[]', () => {
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply ui/main/src/app/services/entities.service.spec.ts]: Done

File: ui/main/src/app/modules/cards/components/detail/detail.component.ts [hunk 5]
```diff
@@ -465,7 +465,9 @@ export class DetailComponent implements OnChanges, OnInit, OnDestroy, AfterViewC
                         });
                 }
             });
-            this._userEntitiesAllowedToRespond = this.card.entitiesAllowedToRespond.filter(x => this.user.entities.includes(x));
+            const entitiesToRespond = this.entitiesService.getEntities().filter(entity  => this.card.entitiesAllowedToRespond.includes(entity.id));
+            const allowed = EntitiesService.getEntitiesAllowedToRespond(entitiesToRespond, this.entitiesService.getEntities()).map(entity => entity.id);
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply ui/main/src/app/modules/cards/components/detail/detail.component.ts]: Done

File: ui/main/src/app/services/entities.service.ts [hunk 6]
```diff
@@ -117,4 +117,25 @@ export class EntitiesService extends CachedCrudService implements OnDestroy {
...
 
+  public static getEntitiesAllowedToRespond(selected: Entity[], allEntities: Entity[]) : Entity[] {
+    let allowed = new Array<Entity>();
+    selected.forEach(entity => {
+        console.log("Foreach entity " + entity.id);
+        if (entity.entityAllowedToSendCard) {
+            console.log("Push entity " + entity.id);
+            allowed.push(entity);
+        } else {
+          console.log("Get childs of entity " + entity.id);
```
[the remark this answers is not in the document]
Comment (quinarygio) [reply ui/main/src/app/services/entities.service.ts]: Done

**Pull request merged**